File: py/1340.py
from queue import Queue
from queue import PriorityQueue


def main():
    while True:
        stack = list()
        queue = Queue()
        priority_queue = PriorityQueue()
        is_stack, is_queue, is_priority_queue = True, True, True

        try:
            n = int(input())
        except EOFError:
            break

        for _ in range(n):
            operation, value = [int(i) for i in input().split()]

            if operation == 1:
                stack.append(value)
                queue.put(value)
                priority_queue.put(-value)
            else:
                if value != stack.pop():
                    is_stack = False
                if value != queue.get():
                    is_queue = False
                if -value != priority_queue.get():
                    is_priority_queue = False

        if ((is_stack and is_queue) or
            (is_stack and is_priority_queue) or
            (is_queue and is_priority_queue)):
            print('not sure')
        elif is_stack or is_queue or is_priority_queue:
            if is_stack:
                print('stack')
            elif is_queue:
                print('queue')
            elif is_priority_queue:
                print('priority queue')
        else:
            print('impossible')


if __name__ == '__main__':
    main()

File: cpp/1340.cpp
#include <iostream>
#include <stack>
#include <queue>

using namespace std;

int main()
{
    int n, operation, value;

    while (cin >> n) {
        stack<int> stack;
        queue<int> queue;
        priority_queue<int> priority_queue;
        bool is_stack = true, is_queue = true, is_priority_queue = true;

        for (n; n > 0; n--) {
            cin >> operation >> value;
            
            if (operation == 1) {
                stack.push(value);
                queue.push(value);
                priority_queue.push(value);
            } else {
                if (value != stack.top()) {
                    is_stack = false;
                }
                if (value != queue.front()) {
                    is_queue = false;
                }
                if (value != priority_queue.top()) {
                    is_priority_queue = false;
                }

                stack.pop();
                queue.pop();
                priority_queue.pop();
            }
        }

        if ((is_stack && is_queue) ||
            (is_stack && is_priority_queue) ||
            (is_queue && is_priority_queue)) {
                cout << "not sure\n";
        } else if (is_stack || is_queue || is_priority_queue) {
            if (is_stack) {
                cout << "stack\n";
            } else if (is_queue) {
                cout << "queue\n";
            } else if (is_priority_queue) {
                cout << "priority queue\n";
            }
        } else {
            cout << "impossible\n";
        }
    }

    return 0;
}